Adjoint and two-adjoint equivalences

daniel-carranza opened the pull request:
This is a revision of the pull request we submitted (under Ryan Sandford) which proves additional results, removes redundant lemmas, and better integrates the built-in equivalence type.

From the previous pull request:
"adj.lean contains definitions and theorems regarding half adjoint equivalences which formalize
4.2.3, 4.2.7, 4.2.9, 4.2.10, 4.2.12, and 4.2.13 from section 4.2 of the HoTT Book.
2adj.lean contains definitions and theorems regarding half 2-adjoint equivalences which formalize analogous theorems to that of 4.2.3, 4.2.12, 4.2.13 from section 4.2 of the HoTT Book."

daniel-carranza committed:
Definitions and theorems about adjoint equivalences and two half-adjoint equivalences.

daniel-carranza committed:
Merge https://github.com/gebner/hott3

daniel-carranza committed:
Updated to include an additional theorem for the
full two-adjoint equivalence; renamed folder to 2_adj

fpvandoorn commented on src/hott/types/2_adj/hty_rec.lean:
```diff
@@ -0,0 +1,55 @@
+
+  -- Based homotopy induction
+  @[hott, induction] def rec {f : Π(x : A), B x} (P : Π(g : Π(x : A), B x), (f ~ g) → Type _) (c : P f (λx, refl (f x)))
+    : Π(g : Π(x : A), B x) (H : f ~ g), P g H :=
```
I think this already exists: https://github.com/gebner/hott3/blob/436a4f5e49f51ca5773b34df046bbf67e2584f70/src/hott/init/funext.lean#L291-L292

fpvandoorn commented on src/hott/types/2_adj/adj.lean:
```diff
@@ -0,0 +1,165 @@
+    is_equiv.inv (is_hadj_l_equiv_is_equiv f) (is_equiv.adjointify f inv right_inv left_inv)
+
+    -- Definition of a full-adjoint equivalence
+    @[hott] def adj (f : A → B) :=
```
It might be useful to give the projections of this definition a name, like 
```
def adj.ε (h : adj f) := h.2.2.1
```
(and same for the other projections) 

If you do that, whenever you have `(h : adj f)` you can write `h.ε` for the corresponding projection.
This should make the file `two_adj.lean` more readable.

fpvandoorn commented on src/hott/types/2_adj/two_adj.lean:
```diff
@@ -0,0 +1,256 @@
+    h.2.2.2.1 (h.1 y) ⬝ (nat_coh f h.1 h.2.2.1 y) = ap02 f (h.2.2.2.2 y)
+
+    -- Definition of a two (right) half-adjoint equivalence
+    @[hott] def is_two_hae (f : A → B) :=
```
It's probably useful to define projections for this (and `is_two_hae_l` / `two_adj` / `is_hadj_l`), too.

fpvandoorn reviewed (commented):
Sorry for being so very late with looking at this PR.

Thanks for using the parts that are already in the library. This is a nice formalization of the properties of 2-adjoint equivalences.
I have a few suggestions you could incorporate, if you want to clean that up (but I'm also fine with merging it as-is).

(You can push new commits directly to your master branch, and this PR will be updated automatically, no need to open a new PR.)

daniel-carranza committed:
Removed duplicate hty.rec, field notation added

daniel-carranza commented:
Went ahead with all the suggestions listed here. I also removed the file hty_rec.lean and moved it into prelim.lean, since it only contained two results after clean-up.

fpvandoorn commented:
Looks good. Thanks for the fixes!
